Makes grounding rods consistent, changes the supermatters bolts a great deal.

LemonInTheDark opened the pull request:
## About The Pull Request


- Changes all occurrences of tesla not solely used by it into zap so things are a little more clear.
- Adds a new bitflag, "ZAP_IS_TESLA".
- Does some touch-ups on my incorrect/misleading comments regarding the supermatter.
- Fixes a small.... bug with the starting power of the sm, it should start at zero rather then one now.
- Fixes a years old bug with the way hallucinations get clamped, clamp() was being misused, which meant a statement was always returning 200
- Fixes this "Bug" #41075 (more on that later).
- Refactored supermatter_zap() so it doesn't have as much repeated code.
- Changed a lot about how the bolts the sm produces interact with the world.
    1. The size of the area the bolts look in is now variable, based on the pressure around the object they are jumping from, and the strength of the bolt. The max is 8, so just over the radius of the screen, and the min is 2, so the bolt can always jump out of the enclosure. 
    2. The limit on how long each total bolt can be is now variable, with the heat of the gas around the sm decreasing it, and the molar count and power increasing it. The function should act like a log function, with the max being 3000 and the min being 350 .
    3. We multiply the power of the bolt touching the tesla coil by 15, so that the power produced is noticeable, check the comment for more details.
    4. Grounding rods and tesla coils do not always draw the bolts, the engine has an erratic behavior to its bolts, I'd like to keep that. 
    5. Bolts like to jump to living things x3 more now, this should keep working near high power dangerous no matter the setup 
    6.A chain of bolts will remember where it has made contact, and will not loop back around 
    7.On paper tesla bolts split less, but thats mostly due to me making static values into calculations, I've basically centered them where I'd like. Please do test it out and let me know if it's too high.
- Fixed a 3 year old bug caused by goof, bicycles are meant to fry when hit with a bolt, but because goof never had the tesla check for them, they never had their zap_act() function called (#24431)
- The tesla and sm will now shoot at vehicles, if they hit, any riders will be zapped.
- It is now takes much more time for one guy to die in a zap storm. I've added a bitflag that keeps track of when someone was last shocked, and sets a timer on it, exactly the same way objects behave.

## Why It's Good For The Game

Adds some consistency to grounding rods and the like, and expands on the options available to the enterprising engineer.

Now for some explanation on how this happened, and why I refactored in the way I did,
The tesla bolts are designed to be static, so if all objects are in the same spots, the tesla will aim for the same things,
The sm was designed to be more erratic, to keep visual interest. So the bolts are randomized in their path, favoring grounded machines over mobs, mobs over machines, and machines over structures.
These two would play nice with each other, except tesla code is not built to do so. zap_act() arc into the next chain in the tesla bolt.

At the time the code was written, the tesla bolts were being integrated with all other lightning effects,
the guy who wrote it wanted to keep the random behavior, and both he and goof either were too lasy, forgot, or had their pc blow up, and so never ended up finishing it. #24785
I intend to fully merge the two functions, but I want to get some opinions on the behavior of the bolts in the mean time, and so I don't spend too much time on something none is interested in.

Coming to that, the only problems I can see with this pr is the state of an overcharged sm, and research points
The sm will produce high amounts of rads, along with pyro anomalies.
This means to be fully protected, you need to be in the ce's hardsuit.
Even through the scaling down that power input receives when the tesla coil is used to generate research, it does produce a significant amount, that needs some looking into.
 
I'd like some feedback on all the formulas I've added, 
Additionally, I'd like to know what you think about upping the damage the bolts can cause, 10 seems a little low to me.

Thank you for your time, I know this is a bit of a dense one.

- [x] Fully refactor zap_act(). (Bad Idea, the way it works currently is the best I can come up with that maintains the preference and random selection behavior)
- [x] Fix the fucked bitflags.
- [x] Produce solid numbers for research production at max operational capacity.
- [x] Extend buckle zap to carrying people, or being carried. Maybe make this depend on insuls.(Already a thing, probably shoulda checked.)

## Changelog
:cl:LemonInTheDark/TheFInalPotato
add: The supermatter has been pulled back down into reality, it will now interact with tesla coils and grounding rods
bug: WEEP cargo mains, the bicycle is no longer a perfect resistor. It will now properly get fryed when zapped by lightning.
add: Vehicles can now be zapped by both the sm and the tesla, they will shock all the people buckled to them 
balance: The supermatter is a lot more touchy when it comes to bolts, it will jump out at you a good amount more now. Be careful
spellcheck: Fixed my old typos
code: Fixed a bunch of small sm related goofs
refactor: Refactored tesla_act into zap_act, it now allows for other zapping methods. Not entirely optimal, but it'll do.
refactor: Refactored supermatter_zap(), it should be a lot cleaner now
/:cl:

LemonInTheDark committed:
Goof I will find you and I will berate you

holy hell

Calm and peaceful ahhhhh

I swear goofs pc failing was the biggest tragedy in gaming

Fixes some flags

More touchups

Now to squash

Goofs PC failing was the biggest tragady in GAMING

LemonInTheDark committed:
Fixing some misspellings before the bee gets me

LemonInTheDark commented:
It looks like the extensive variable use in zap() is to lower the amount of typecasting, I'm debating whether the cleanliness is worth the extra function call

LemonInTheDark commented:
Ok, so the way the tesla denotes hitting something is a bitflag used by obj. but since we don't wanna get ganged up on by bolts, I need to refactor it to be an atom thing I think? Thoughts/suggestions would be much appreciated.

LemonInTheDark committed:
GOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF

LemonInTheDark commented:
I'm also gonna make tesla's target selection a bit more readable, gonna try making a list for it to iterate through, one for each type we want to discriminate, conflicts resolved with distance, then have it start at the top and find the first non null entry.

LemonInTheDark committed:
Full vehicle support

LemonInTheDark committed:
Changes some flags around, "fixes" some bugs, makes being shocked less insane

LemonInTheDark commented:
Research point increase seems quite reasonable, 3 coils with maxed capacitors running at 7000 power produced 60 points over a minute.
[Before](https://cdn.discordapp.com/attachments/458452245256601615/663253530097876992/unknown.pngl)
[After](https://cdn.discordapp.com/attachments/458452245256601615/663253728375210003/unknown.pngl)

LemonInTheDark committed:
Makes spliting more reasonable, unfucks some shitty tesla code I copypasta'd into the sm, godspeed shitcode,
none noticed you.

LemonInTheDark commented:
I've appropriated TESLA_IGNORE_1's bitmask spot and turned it into IS_SHOCKED, which keeps track of shocks and shit. The way I've done this removes the shock immunity from lightning spells and the reactive armor, see main post for more details.

LemonInTheDark commented:
This mess is "_ready_" for review, let me know what you think.

ShizCalev commented on code/modules/clothing/suits/reactive_armour.dm:
```diff
@@ -151,19 +151,9 @@
-/obj/item/clothing/suit/armor/reactive/tesla/equipped(mob/user, slot)
-	..()
-	if(slot_flags & slot) //Was equipped to a valid slot for this item?
-		user.flags_1 |= TESLA_IGNORE_1
```
This isn't really fixing the issue, it's just sweeping it under the rug until something else gets added that causes it.

LemonInTheDark replied:
That's true. It is kinda a bandage fix. Would you like me to make it a trait, or is that too much of a operation to add to Tesla code? I could make it a bitflag tied to a trait if checking has_trait is too expensive.

LemonInTheDark replied:
I've re-added the relevant code, I await the impending merge conflict.

LemonInTheDark committed:
Removes my tesla_ignore_1 fix, another pr is up that should solve things, so we're just leaving it in for now,
gonna have to deal with merge conflicts when it gets merged, but that won't be too bad

LemonInTheDark committed:
tfw

LemonInTheDark committed:
Nothing to see here folks

kriskog commented:
Merge conflict

LemonInTheDark committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into zap_zap

LemonInTheDark committed:
I'm a boomer

JAremko commented:
# epic :scream_cat: 

> Bolts like to jump to living things x3 more now, this should keep working near high power dangerous no matter the setup

Would tesla shocks destroy equipment like freezers, pipe controls and SM APC?   Even something like randomising machine settings when zapped (messing freezer target temperature, switching pipes, turning equipment of light if APC zapped) would make engie life much more spicy :smile:  

Even turning you into gibs would be OK if there was a way to make small areas lightning prove.  Like if 3X3 square around grounding roads had 100% protection from zaps :thinking:  I mean something like this can be brute force by feeling entire room with grounding road :crossed_fingers: and praying to RNGesus. But its kinda dumb :thinking:

JAremko commented:
> Vehicles can now be zapped by both the sm and the tesla, they will shock all the people buckled to them

@LemonInTheDark  Does it work with cluwn car. Hope so  :thinking:

LemonInTheDark commented:
> Would tesla shocks destroy equipment like freezers, pipe controls and SM APC? Even something like randomising machine settings when zapped (messing freezer target temperature, switching pipes, turning equipment of light if APC zapped) would make engie life much more spicy 😄

The tesla does things like that, I wanted to avoid that so the engine could be used for power without hellish risks, and so power delams aren't impossible to stop.
I considered a smaller emp effect, but never did anything like it. Feel free to try it, it's just changing a few flags

> Even turning you into gibs would be OK if there was a way to make small areas lightning proof. Like if 3X3 square around grounding roads had 100% protection from zaps 🤔 I mean something like this can be brute forced by filling entire room with grounding road 🤞 and praying to RNGesus. But its kinda dumb 🤔 What's cool about SM is that you can build extremely robust setup but a single mistake can cost you dearly. So there always is "the pie in the sky" - theoretically perfect SM setup. But if you can get F-ed by a random lightning every 101's run its kinda meh. And if a lightning hasn't ability to F you at all it is also meh 😄

I agree, I don't like how bolts get completely stopped by gloves, but I don't like how powerful the tesla shocks are either, might need some more tweaking.
About the jumping, I tried it with the rods having no ignore chance, but it was way too manageable.
The chance should be low enough to be mitigated easily.

JAremko commented:
> I tried it with the rods having no ignore chance, but it was way too manageable.
The chance should be low enough to be mitigated easily.

@LaharlMontogmmery  what if  grounding rods needed some babysitting instead :thinking:  For example, grounding rod can store some charge (similar to field generators) that slowly dissipates. The better capacitor it has - the bigger storage amount. Sometimes capacitors randomly fails (like a light bulb) and need to be replaced. So the more grounding rods you have the more annoying it gets. 
When a grounding rod gets overloaded it violently explodes/starts to zap everything. To prevent it you can reset the rods by swapping capacitors directly or by rebuilding the rod (similar to how we clean freezers/heaters from waste gases) it  will also mean that you gonna need pathing around them.

To make  engie lives even shorter rod wrench on top of a wire should zap (like it used to?) - it will make carefuller planing necessary and any rod with some charge in the capacitor will electrocute you on contact. More happy accidents!
